File: internal/notify/notify_test.go
package notify

import (
	"context"
	"encoding/json"
	"io"
	"net/http"
	"net/http/httptest"
	"testing"
)

// captureServer records the body of the next POST it receives.
func captureServer(t *testing.T, status int) (*httptest.Server, *[]byte) {
	t.Helper()
	var got []byte
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		b, _ := io.ReadAll(r.Body)
		got = b
		w.WriteHeader(status)
	}))
	t.Cleanup(srv.Close)
	return srv, &got
}

func TestSlackNotify(t *testing.T) {
	srv, body := captureServer(t, http.StatusOK)
	n := NewSlack(srv.URL)

	if err := n.Notify(context.Background(), Message{Title: "Digest", Text: "3 commits", Link: "https://x/pr/1"}); err != nil {
		t.Fatalf("Notify: %v", err)
	}
	var payload map[string]string
	if err := json.Unmarshal(*body, &payload); err != nil {
		t.Fatalf("unmarshal: %v", err)
	}
	want := "*Digest*\n3 commits\n<https://x/pr/1>"
	if payload["text"] != want {
		t.Errorf("text = %q, want %q", payload["text"], want)
	}
}

func TestTeamsNotify(t *testing.T) {
	srv, body := captureServer(t, http.StatusOK)
	n := NewTeams(srv.URL)

	if err := n.Notify(context.Background(), Message{Title: "Result", Text: "fixed", Link: "https://x/pr/2"}); err != nil {
		t.Fatalf("Notify: %v", err)
	}
	var payload map[string]any
	if err := json.Unmarshal(*body, &payload); err != nil {
		t.Fatalf("unmarshal: %v", err)
	}
	if payload["type"] != "message" {
		t.Errorf("type = %v, want message", payload["type"])
	}
	atts, ok := payload["attachments"].([]any)
	if !ok || len(atts) != 1 {
		t.Fatalf("attachments = %v", payload["attachments"])
	}
	att := atts[0].(map[string]any)
	if att["contentType"] != "application/vnd.microsoft.card.adaptive" {
		t.Errorf("contentType = %v", att["contentType"])
	}
	content := att["content"].(map[string]any)
	if content["type"] != "AdaptiveCard" {
		t.Errorf("content type = %v", content["type"])
	}
	if _, hasActions := content["actions"]; !hasActions {
		t.Error("expected actions for a message with a link")
	}
}

func TestNon2xxIsError(t *testing.T) {
	srv, _ := captureServer(t, http.StatusInternalServerError)
	if err := NewSlack(srv.URL).Notify(context.Background(), Message{Text: "x"}); err == nil {
		t.Fatal("expected error on 500")
	}
}

func TestNewFactory(t *testing.T) {
	if _, err := New("slack", "https://hook", ""); err != nil {
		t.Errorf("slack: %v", err)
	}
	if _, err := New("teams", "", "https://hook"); err != nil {
		t.Errorf("teams: %v", err)
	}
	if _, err := New("slack", "", ""); err == nil {
		t.Error("slack without URL should error")
	}
	if _, err := New("discord", "a", "b"); err == nil {
		t.Error("unknown provider should error")
	}
}

func TestTeamsCardOmitsActionsWithoutLink(t *testing.T) {
	card := teamsCard(Message{Title: "t", Text: "b"})
	content := card["attachments"].([]map[string]any)[0]["content"].(map[string]any)
	if _, has := content["actions"]; has {
		t.Error("no link -> no actions")
	}
}

File: kotlin/src/test/kotlin/com/automation/agent/notify/NotifyTest.kt
package com.automation.agent.notify

import com.sun.net.httpserver.HttpServer
import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.spec.style.BehaviorSpec
import io.kotest.matchers.shouldBe
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import java.io.IOException
import java.net.InetSocketAddress

private class Capture {
    @Volatile
    var body: ByteArray = ByteArray(0)
}

/** Starts a local server that records the body of the next POST and replies with [status]. */
private fun captureServer(status: Int): Pair<HttpServer, Capture> {
    val cap = Capture()
    val server = HttpServer.create(InetSocketAddress("127.0.0.1", 0), 0)
    server.createContext("/") { exchange ->
        cap.body = exchange.requestBody.readBytes()
        exchange.sendResponseHeaders(status, -1)
        exchange.close()
    }
    server.start()
    return server to cap
}

private fun HttpServer.url(): String = "http://127.0.0.1:${address.port}/"

class NotifyTest : BehaviorSpec({
    Given("a Slack notifier and a capturing webhook") {
        When("notifying a message with title, text and link") {
            Then("the payload is rendered as Slack mrkdwn") {
                val (srv, cap) = captureServer(200)
                try {
                    newSlack(srv.url())
                        .notify(Message(title = "Digest", text = "3 commits", link = "https://x/pr/1"))
                    val payload = Json.parseToJsonElement(String(cap.body)).jsonObject
                    payload.getValue("text").jsonPrimitive.content shouldBe "*Digest*\n3 commits\n<https://x/pr/1>"
                } finally {
                    srv.stop(0)
                }
            }
        }
    }

    Given("a Teams notifier and a capturing webhook") {
        When("notifying a message with a link") {
            Then("it posts a Workflows adaptive card with an action") {
                val (srv, cap) = captureServer(200)
                try {
                    newTeams(srv.url())
                        .notify(Message(title = "Result", text = "fixed", link = "https://x/pr/2"))
                    val payload = Json.parseToJsonElement(String(cap.body)).jsonObject
                    payload.getValue("type").jsonPrimitive.content shouldBe "message"
                    val atts = payload.getValue("attachments").jsonArray
                    atts.size shouldBe 1
                    val att = atts[0].jsonObject
                    att.getValue("contentType").jsonPrimitive.content shouldBe
                        "application/vnd.microsoft.card.adaptive"
                    val content = att.getValue("content").jsonObject
                    content.getValue("type").jsonPrimitive.content shouldBe "AdaptiveCard"
                    content.containsKey("actions") shouldBe true
                } finally {
                    srv.stop(0)
                }
            }
        }
    }

    Given("a webhook that returns a 500 status") {
        When("notifying via Slack") {
            Then("it raises an IOException") {
                val (srv, _) = captureServer(500)
                try {
                    shouldThrow<IOException> {
                        newSlack(srv.url()).notify(Message(text = "x"))
                    }
                } finally {
                    srv.stop(0)
                }
            }
        }
    }

    Given("the notifier factory") {
        When("given valid providers and matching urls") {
            Then("it builds a notifier for each") {
                newNotifier("slack", "https://hook", "")
                newNotifier("teams", "", "https://hook")
            }
        }
        When("given a missing url or an unknown provider") {
            Then("it fails") {
                shouldThrow<IllegalArgumentException> { newNotifier("slack", "", "") }
                shouldThrow<IllegalArgumentException> { newNotifier("discord", "a", "b") }
            }
        }
    }

    Given("a Teams message without a link") {
        When("building the adaptive card") {
            Then("it omits the actions block") {
                val card = teamsCard(Message(title = "t", text = "b"))
                val content = card.getValue("attachments").jsonArray[0].jsonObject.getValue("content").jsonObject
                content.containsKey("actions") shouldBe false
            }
        }
    }
})
